File: autocomplete_system_2.py
from collections import defaultdict, deque
import heapq
from typing import List


class AutocompleteSystem:
    class Pair:
        def __init__(self, sentence, count):
            self.sentence = sentence
            self.count = count

        def __lt__(self, other):
            if self.count == other.count:
                return self.sentence > other.sentence
            return self.count < other.count

    def __init__(self, sentences: List[str], times: List[int]):
        self.dic = defaultdict(int)
        for i in range(len(sentences)):
            self.dic[sentences[i]] += times[i]
        self.search = ''

    def input(self, c: str) -> List[str]:
        pq = []
        if c == '#':
            self.dic[self.search] += 1
            self.search = ''
            return pq
        self.search += c
        for k, v in self.dic.items():
            if k.startswith(self.search):
                heapq.heappush(pq, self.Pair(k, v))
            if len(pq) > 3:
                heapq.heappop(pq)
        result = deque()
        while pq:
            count_sentence = heapq.heappop(pq)
            result.insert(0, count_sentence.sentence)
        return result

File: AutocompleteSystem2.java
import java.util.*;

class AutocompleteSystem2 {
    StringBuilder input;
    class TrieNode{
        HashMap<String, Integer> map;
        HashMap<Character, TrieNode> children;
        public TrieNode(){
            map = new HashMap<>();
            children = new HashMap<>();
        }
    }
    TrieNode root;
    private void insert(String word, int times){
        TrieNode curr = root;
        for(int i = 0; i < word.length(); i++){
            char c = word.charAt(i);
            if(!curr.children.containsKey(c)){
                curr.children.put(c, new TrieNode());
            }
            curr = curr.children.get(c);
            curr.map.put(word, curr.map.getOrDefault(word, 0) + times); 
        }
    }
    private HashMap<String, Integer> search(String word){
        TrieNode curr = root;
        for(int i = 0; i < word.length(); i++){
            char c = word.charAt(i);
            if(!curr.children.containsKey(c)){
                return new HashMap<>();
            }
            curr = curr.children.get(c);
        }
        return curr.map;
    }
    
    public AutocompleteSystem2(String[] sentences, int[] times) {
        root = new TrieNode();
        input = new StringBuilder();
        for(int i = 0; i < sentences.length; i++){
            insert(sentences[i],times[i]);
        }
    }
    
    public List<String> input(char c) {
        if(c == '#'){
            String in = input.toString();
            insert(in, 1);
            input = new StringBuilder();
            return new ArrayList<>();
        }
        input.append(c);
        HashMap<String, Integer> count = search(input.toString());

        PriorityQueue<String> pq = new PriorityQueue<>((a,b)->{
            if(count.get(a) == count.get(b)){
                return b.compareTo(a);
            } 
            return count.get(a) - count.get(b);
        });
        for(String key: count.keySet()){
            String in = input.toString();
            if(key.startsWith(in)){
                pq.add(key);
                if(pq.size() > 3){
                    pq.poll();
                }
            }
        }
        List<String> result = new ArrayList<>();
        while(!pq.isEmpty()){
            result.add(0, pq.poll());

        }
        return result;
    }
}
    

/**
 * Your AutocompleteSystem object will be instantiated and called as such:
 * AutocompleteSystem obj = new AutocompleteSystem(sentences, times);
 * List<String> param_1 = obj.input(c);
 */